biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotations.java

```diff
@@ -26,55 +48,123 @@ public boolean analyzeJar(Analyzer analyzer) throws Exception {
+		TreeSet<String> requires = new TreeSet<String>();
+		Version maxVersion = AnnotationReader.V1_0;
+
 		for (Clazz c : list) {
 			for (Instruction instruction : instructions.keySet()) {
 
 				if (instruction.matches(c.getFQN())) {
 					if (instruction.isNegated())
 						break;
-					ComponentDef definition = AnnotationReader.getDefinition(c, analyzer);
+					ComponentDef definition = AnnotationReader.getDefinition(c, analyzer, flags);
 					if (definition != null) {
 
 						definition.sortReferences();
 						definition.prepare(analyzer);
-						verifyVersion(analyzer, definition);
 
 						String name = "OSGI-INF/"
 								+ analyzer.validResourcePath(definition.name, "Invalid component name") + ".xml";
 						names.add(name);
 						analyzer.getJar().putResource(name, new TagResource(definition.getTag()));
+
+						if (definition.service != null) {
+							String[] objectClass = new String[definition.service.length];
+
+							for (int i = 0; i < definition.service.length; i++) {
+								Descriptors.TypeRef tr = definition.service[i];
+								objectClass[i] = tr.getFQN();
+							}
+							Arrays.sort(objectClass);
+							addServiceCapability(objectClass, provides);
+						}
+						for (ReferenceDef ref : definition.references.values()) {
+							String objectClass = ref.service;
+							addServiceRequirement(objectClass, requires);
+						}
+						maxVersion = ComponentDef.max(maxVersion, definition.version);
 					}
 				}
 			}
 		}
+		if (flags.contains(Flag.extender)
+				|| maxVersion.compareTo(AnnotationReader.V1_3) >= 0) {
+			addExtenderRequirement(requires);
+		}
 		sc = Processor.append(names.toArray(new String[names.size()]));
 		analyzer.setProperty(Constants.SERVICE_COMPONENT, sc);
+		updateHeader(analyzer, Constants.REQUIRE_CAPABILITY, requires);
+		updateHeader(analyzer, Constants.PROVIDE_CAPABILITY, provides);
 		return false;
 	}
 
+	private void addServiceCapability(String[] objectClass, Set<String> provides) {
+		if (objectClass.length > 0) {
+			Parameters p = new Parameters();
+			Attrs a = new Attrs();
+			StringBuilder sb = new StringBuilder();
+			String sep = "";
+			for (String oc : objectClass) {
+				sb.append(sep).append(oc);
+				sep = ",";
+			}
+			a.put("objectClass:List<String>", sb.toString());
+			p.put("osgi.service", a);
+			String s = p.toString();
+			provides.add(s);
+		}
+	}
+
+	private void addServiceRequirement(String objectClass, Set<String> requires) {
+		Parameters p = new Parameters();
+		Attrs a = new Attrs();
+		a.put(Constants.FILTER_DIRECTIVE, "\"(objectClass=" + objectClass + ")\"");
+		a.put(Constants.EFFECTIVE_DIRECTIVE, "\"active\"");
+		p.put("osgi.service", a);
+		String s = p.toString();
+		requires.add(s);
+	}
+
+	private void addExtenderRequirement(Set<String> requires) {
+		Parameters p = new Parameters();
+		Attrs a = new Attrs();
+		a.put(Constants.FILTER_DIRECTIVE, "\"(&(osgi.extender=osgi.component)(version>=1.3)(!(version>=2.0)))\"");
```

Comment:
Shouldn't the version (1.3) be based upon an argument? Like maxVersion from line 92? Also the upper bound (2.0) should also be computed from the version argument. This prepares us for 1.4+.